### DISCUSSION TITLE
Custom async remote certificate validation with HttpClient

### DISCUSSION THREAD

**Discussion by mddddb** [dotnet/runtime#78761] [Q&A]
We want to have custom remote/server certificate validation for the endpoint we are connecting to using HttpClient.
Currently we can do this by using one of the two callbacks below:
https://github.com/dotnet/runtime/blob/4cbe6f99d23e04c56a89251d49de1b0f14000427/src/libraries/System.Net.Http/src/System/Net/Http/HttpClientHandler.cs#L261
https://github.com/dotnet/runtime/blob/a5f3676cc71e176084f0f7f1f6beeecd86fbeafc/src/libraries/System.Net.Security/src/System/Net/Security/SslClientAuthenticationOptions.cs#L26

```c#
services
    .AddHttpClient("someEndpointClient")
    .ConfigureHttpMessageHandlerBuilder(builder =>
    {
        var handler = (HttpClientHandler)builder.PrimaryHandler;
        handler.ServerCertificateCustomValidationCallback = (httpRequestMessage, certificate, chain, sslPolicyErrors) =>
        {
            // do the actual validation

            return true;
        };
    });
```

The issue is that we have async certificate validation, and having an issue with how we can attach our async validation here without sync over async.

Is there any way to achieve this properly? What options do we have here?
Is there any way to attach our validation somewhere after ServerCertificateCustomValidationCallback, but before the actual request is being sent to the endpoint? I.e. can we add something async in HttpClient after the serverCertificate is available, but before sending the request? Right after the SSL handshake

**Comment by karelz** [comment]
@wfurt @rzikm I believe we avoided async due to underlying platforms not supporting it. Or do I misremember?
If I am right, then sync over async is the best approach.

cc @dotnet/ncl

**Reply by rzikm** [reply]
From looking at the code, I think we can technically add and implement async callbacks both for certificate selection and validation. For SslStream, neither of these callbacks are invoked from native code. They are, however, invoked from native code in `QuicConnection`, moving them elsewhere would require support from MsQuic (which we already asked for IIRC).

Maybe @wfurt remembers something I don't know.

**Reply by mddddb** [reply]
@rzikm Just to clarify, technically it is possible to make the changes in `SslStream`, but under that we use this `QuicConnection`?
And `QuicConnection` is not owned by you?
Would this (https://source.dot.net/#System.Net.Quic/System/Net/Quic/QuicConnection.SslConnectionOptions.cs,129) be the place where the change is needed?

And you were mentioning an ask from this `QuicConnection`'s devs, do you mean that you have already asked them to support async in there?

**Reply by ManickaP** [reply]
`QuicConnection` is completely separate from `SslStream`, it just uses the same option class so that users of both can have the same way to configure them. And it is owned by us. 

However, adding async callbacks would mean a new API surface. We cannot just change the existing ones. We would also need to support both at the same and decide on precedence that would cause the least amount of confusion. And it would not be sooner than .NET 8 time frame, as we never back-port new APIs. 
What I'm trying to say is that even if this is technically possible to implement, it might not be the best possible option to do so.

**Reply by rzikm** [reply]
`QuicConnection` is a separate class that implements the QUIC protocol which also uses TLS for encryption. Both `QuicConnection` and `SslStream` are maintained by our team. However, both `SslStream` and `QuicConnection` accepts `Ssl(Client|Server)AuthenticationOptions` instance for configuration. If we add new properties to the auth options, we should support them in both places.

`QuicConnection` is currently a wrapper over the [MsQuic](https://github.com/microsoft/msquic) native library which is maintained by a different team. Since MsQuic programming model gives us callbacks with events, we would be currently forced to block/wait on the async calls from inside the callback thread, which could negatively affect the connection (and even other, already established connections).

The issue with long-running callbacks exists for synchronous callbacks as well, which is something we would like to fix anyway, along with other feature gaps that exist when compared to `SslStream`

**Reply by simonrozsival** [reply]
The validation callbacks in `AndroidMessageHandler` and `NSUrlSessionHandler` are invoked from native code and I'm not sure if it would be possible to implement async validation on those platforms without blocking. Also when #77386 is finalized and merged, the validation callback in the SslStream will be invoked from native code on Android.

**Reply by wfurt** [reply]
We may change that on other platforms in future. Doing validation after handshake has some behavior differences. It still may be possible but we would need to find some way of synchronization. The current `sync` version just works.

**Comment by ManickaP** [comment]
> Is there any way to attach our validation somewhere after ServerCertificateCustomValidationCallback, but before the actual request is being sent to the endpoint? 

If you need something after TLS handshake and before the first request then [`SocketsHttpHandler.PlaintextStreamFilter`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.plaintextstreamfilter?view=net-7.0) might help. This is called after the TLS handshake finished, so you cannot refuse the cert per se. But you can throw an exception from it, which will close the connection and fail the request that started that connection.

Other option would be [`SocketsHttpHandler.ConnectCallback`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-7.0) where you have a full control over the connection creation. In this callback you can also establish TLS on top of the socket and do your thing to validate the cert. But this is much more involved.

**Reply by mddddb** [reply]
Thank you, @ManickaP! [SocketsHttpHandler.PlaintextStreamFilter](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.plaintextstreamfilter?view=net-7.0) might be able to serve as a workaround for us. I will do some experiments with that one and see whether it is suitable for us

**Reply by mddddb** [reply]
@ManickaP Seems like we might be able to work around using [SocketsHttpHandler.PlaintextStreamFilter](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.plaintextstreamfilter?view=net-7.0), still continuing with the research though.
One thing, might have missed it, but in case of QuicConnection instead of SslStream (HTTP3???) I didn't find the place where `PlainTextStreamFilter` gets called. I guess you couldn't even call it even if you needed I guess, since it's not a stream.
I just wanted to confirm if the workaround is only for SslStream and not QuicConnection

**Reply by ManickaP** [reply]
It is only for `SslStream` thus for HTTP/1.1 and HTTP/2. There's no such thing for `QuicConnection` so this won't work for HTTP/3. If you don't mind not being able to use HTTP/3, you can set up DefaultHttpVersion and DefaultVersionPolicy for your requests to prevent upgrades to HTTP/3.

**Reply by mddddb** [reply]
@ManickaP Thanks for clarifying🙂

**Comment by wfurt** [comment]
I think this is mostly Historical. With the APM the pattern would be pretty hard to use so we started with Sync versions for both selection & validation. We added Async version of selection few releases back but the need for async validation was minimal so far. 

If you can live with HTTP/1.1 and 2 (e.g. don't need Quic) the `ConnectCallback` @ManickaP  suggested is probably best option. 
That needs 7.0 (#63851) and it will not work in 6.0 or prior.

**Comment by mddddb** [comment]
What is the plan for future versions? For .NET8?
Is there a plan to make the callback async? Can we request such support in the upcoming versions?
@ManickaP, you were mentioning something around .NET8 release timelines, was that a theoretical ETA if you decide to add the support now? Or is it an ETA for a feature that already is in the plans?

**Reply by ManickaP** [reply]
Theoretical, in other words, the soonest possible time frame if we decide to do anything.
If you wish something like this, you can file an API suggestion issue in this repo. You can follow this documentation about our API review process: https://github.com/dotnet/runtime/blob/734cd446f888af04656e779a6f9b6e1c6c306aaf/docs/project/api-review-process.md

Then, we'd discuss it in the team and decide, whether to do it and if so when.

EDIT: as this is @rzikm  and @wfurt area, they might already have some ideas and thoughts on the future of this.

**Reply by wfurt** [reply]
I think the next step would be API proposal. And investigation if it is feasible.  General ask for `async`  make sense to me.

**Reply by mddddb** [reply]
@wfurt Will do, haven't read the guidelines on API proposals yet. Will create an API proposal next week

**Reply by rymeskar** [reply]
@wfurt , @ManickaP is the ask for my team to file the API proposal as well as go through the API review process? Or we can only file the async requirements and goals and you can take it from there? 

I am sorry, I am not sure what is the usual process here.

**Reply by CarnaViire** [reply]
> Or we can only file the async requirements and goals and you can take it from there?

Yes, you only need to file the issue

**Reply by mddddb** [reply]
@wfurt @ManickaP @CarnaViire https://github.com/dotnet/runtime/issues/79441 was opened. Please, let us know if anything else is needed from us at this stage
